### PR TITLE
feat: vectorize select

### PR DESCRIPTION
**Pull request by polvalente**

### CODE REVIEW

File: nx/lib/nx.ex [hunk 1]
```diff
@@ -6365,28 +6365,101 @@ defmodule Nx do
...
+          ],
+          [
+            [1, 1, 2],
+            [1, 1, 3]
+          ]
+        ]
+      >
+
+    In the next example, notice that even though the `pred` input
+    is scalar, because we're dealing with vectorize inputs, some
```
Comment (josevalim) [inline nx/lib/nx.ex:6393]:
```suggestion
    is scalar, because we're dealing with vectorized inputs, some
```

File: nx/lib/nx.ex [hunk 2]
```diff
@@ -6365,28 +6365,101 @@ defmodule Nx do
...
 
     {output_shape, output_names} =
-      case pred_shape do
-        {} ->
+      case {vectorized_axes, pred_shape} do
+        {[], {}} ->
           Nx.Shape.binary_broadcast(true_shape, true_names, false_shape, false_names)
 
         _ ->
-          {pred_shape, pred_names}
```
Comment (josevalim) [inline nx/lib/nx.ex:6387]: Should we still match on `{[], _}` and return `{pred_shape, pred_names}`?

File: nx/lib/nx.ex [hunk 3]
```diff
@@ -6401,8 +6474,36 @@ defmodule Nx do
         Nx.Shape.broadcast_axes(false_shape, output_shape)
       )
 
-    out = %{pred | shape: output_shape, type: output_type, names: output_names}
-    impl!(pred, on_true, on_false).select(out, pred, on_true, on_false)
+    out = devectorize(%{pred | shape: output_shape, type: output_type, names: output_names})
+
+    pred = devectorize(pred)
```
Comment (josevalim) [inline nx/lib/nx.ex:6479]: Maybe invert the order of these two operations so you don't have to devectorize `out`?

Comment (josevalim) [inline nx/lib/nx.ex:6479]: Maybe invert the order of these two operations so you don't have to devectorize `out`?

```suggestion
    pred = devectorize(pred)
    out = %{pred | shape: output_shape, type: output_type, names: output_names}
```

?

**Review by josevalim — APPROVED** [review]

File: nx/lib/nx.ex (hunk 2 above)
Comment (polvalente) [reply nx/lib/nx.ex:6387]: Makes sense. Avoids some work :)

File: nx/lib/nx.ex (hunk 3 above)
Comment (polvalente) [reply nx/lib/nx.ex:6479]: We'd have to recalculate `output_shape` if we invert the order

**Pull request merged**